File: osm.js
import {distance} from './geometry.js';
export function classify(tags={}){
 const s=String(tags.surface||'').toLowerCase().trim();
 if(s==='asphalt')return {kind:'asphalt',source:'osm'};
 if(['gravel','fine_gravel','compacted','pebblestone'].includes(s))return {kind:'gravel',source:'osm'};
 if(['concrete','concrete:plates','concrete:lanes','paving_stones','sett','cobblestone','wood'].includes(s))return {kind:'gravel',source:'osm'};
 if(s==='paved')return {kind:'gravel',source:'osm'};
 if(['path','bridleway','footway'].includes(tags.highway)&&(!s||['ground','dirt','earth','grass','mud','rock','unpaved','sand'].includes(s)))return {kind:'trail',source:s?'osm':'osm-inferred'};
 if(['unpaved','ground','dirt','earth','grass','mud','sand'].includes(s))return {kind:'gravel',source:'osm'};
 if(!s&&tags.tracktype==='grade2')return {kind:'gravel',source:'osm-inferred'};
 return {kind:'unknown',source:'unknown'};
}
export function infer(points,elements){
 const lat0=points.reduce((s,p)=>s+p[0],0)/points.length,sx=111320*Math.max(.1,Math.cos(lat0*Math.PI/180));
 const xy=p=>[p[1]*sx,p[0]*111320];const grid=new Map();let refs=0,geometryWays=0,taggedWays=0;
 for(const way of elements){const geo=way.geometry||[];if(geo.length<2)continue;geometryWays++;const cls=classify(way.tags);if(cls.kind!=='unknown')taggedWays++;
  for(let i=1;i<geo.length;i++){const a=geo[i-1],b=geo[i];if(!a||!b)continue;const [ax,ay]=xy([a.lat,a.lon]),[bx,by]=xy([b.lat,b.lon]);if(Math.hypot(bx-ax,by-ay)>2000)continue;
   const edge={ax,ay,bx,by,...cls};for(let x=Math.floor(Math.min(ax,bx)/60);x<=Math.floor(Math.max(ax,bx)/60);x++)for(let y=Math.floor(Math.min(ay,by)/60);y<=Math.floor(Math.max(ay,by)/60);y++){const key=x+','+y;if(!grid.has(key))grid.set(key,[]);grid.get(key).push(edge);if(++refs>800000)throw Error('Kartområdet är för stort för matchning. Markera underlag manuellt.');}
  }
 }
 if(elements.length&&!geometryWays)throw Error('Karttjänsten returnerade vägar utan geometri. Inga underlag ändrades.');
 const spans=[];let near=0,known=0;
 for(let i=1;i<points.length;i++){const a=points[i-1],b=points[i];if(a[4]!==b[4]||b[3]<=a[3])continue;const n=Math.max(1,Math.ceil((b[3]-a[3])/25));
  for(let j=0;j<n;j++){const t=(j+.5)/n,[x,y]=xy([a[0]+(b[0]-a[0])*t,a[1]+(b[1]-a[1])*t]),candidates=new Map();const gx=Math.floor(x/60),gy=Math.floor(y/60);
   for(let dx=-1;dx<=1;dx++)for(let dy=-1;dy<=1;dy++)for(const e of grid.get((gx+dx)+','+(gy+dy))||[]){const vx=e.bx-e.ax,vy=e.by-e.ay,f=Math.max(0,Math.min(1,((x-e.ax)*vx+(y-e.ay)*vy)/Math.max(.0001,vx*vx+vy*vy))),d=Math.hypot(x-e.ax-f*vx,y-e.ay-f*vy);if(d<=30&&d<(candidates.get(e.kind)?.d??Infinity))candidates.set(e.kind,{...e,d});}
   const ranked=[...candidates.values()].sort((a,b)=>a.d-b.d);const len=(b[3]-a[3])/n;if(ranked.length)near+=len;
   const c=ranked.length&&(ranked.length===1||ranked[1].d-ranked[0].d>5)?ranked[0]:{kind:'unknown',source:'unknown'};
   if(c.kind!=='unknown')known+=len;const start=a[3]+len*j,end=a[3]+len*(j+1),last=spans.at(-1);if(last&&last.kind===c.kind&&last.source===c.source)last.end=end;else spans.push({start,end,kind:c.kind,source:c.source});
  }
 }
 return {spans,diagnostics:{ways:elements.length,geometry_ways:geometryWays,classified_ways:taggedWays,matched_percent:Math.round(near/points.at(-1)[3]*100),classified_percent:Math.round(known/points.at(-1)[3]*100),checked_at:new Date().toISOString()}};
}
export async function fetchSuggestions(points,fetcher=fetch){
 const south=Math.min(...points.map(p=>p[0]))-.001,north=Math.max(...points.map(p=>p[0]))+.001,west=Math.min(...points.map(p=>p[1]))-.002,east=Math.max(...points.map(p=>p[1]))+.002;
 const area=distance([south,west],[north,west])*distance([south,west],[south,east])/1e6;
 if(area>1200||points.at(-1)[3]>200000)throw Error('Kartförslag stöder högst 200 km och ett kartområde på 1 200 km². Markera denna rutt manuellt.');
 // Do not filter out roads without surface: path/track tags can add information,
 // and unknown parallel roads must participate to avoid false assignments.
 const query=`[out:json][timeout:30][maxsize:33554432];way["highway"](${south},${west},${north},${east});out body geom;`;
 let data;
 try{const res=await fetcher('https://overpass-api.de/api/interpreter',{method:'POST',body:new URLSearchParams({data:query}),headers:{'User-Agent':'AlvinsRuttbank/2.0'},signal:AbortSignal.timeout(45000)});if(!res.ok)throw Error();const reader=res.body.getReader();let bytes=0;const chunks=[];while(true){const {value,done}=await reader.read();if(done)break;bytes+=value.length;if(bytes>16*1024*1024){await reader.cancel();throw Error();}chunks.push(value);}data=JSON.parse(Buffer.concat(chunks).toString('utf8'));if(data.remark||!Array.isArray(data.elements))throw Error();}
 catch{throw Error('Karttjänsten kunde inte svara just nu. Inga underlag ändrades. Försök senare.');}
 return infer(points,data.elements);
}

File: osm.py
"""Optional conservative surface suggestions from OpenStreetMap / Overpass."""
import json
import math
from urllib.request import Request, urlopen
from urllib.parse import urlencode
from geometry import distance

def kind(tags):
    surface=tags.get('surface','')
    if surface=='asphalt':return 'asphalt'
    if tags.get('highway') in ('path','bridleway') and surface in ('ground','dirt','earth','grass','mud','rock'):
        return 'trail'
    if surface in ('gravel','fine_gravel','compacted','pebblestone'):return 'gravel'
    if surface in ('concrete','concrete:plates','paving_stones','sett','cobblestone','wood'):return 'other'
    return 'unknown'

def infer(points, elements):
    lat0=sum(p[0] for p in points)/len(points)
    sx=111320*max(.1,math.cos(math.radians(lat0)))
    def xy(p):return p[1]*sx,p[0]*111320
    grid={};edge_count=0
    for way in elements:
        geo=way.get('geometry',[]);k=kind(way.get('tags',{}))
        if k=='unknown':continue
        for a,b in zip(geo,geo[1:]):
            ax,ay=xy([a['lat'],a['lon']]);bx,by=xy([b['lat'],b['lon']])
            if math.hypot(bx-ax,by-ay)>1000:continue
            edge=(ax,ay,bx,by,k)
            for x in range(math.floor(min(ax,bx)/50),math.floor(max(ax,bx)/50)+1):
                for y in range(math.floor(min(ay,by)/50),math.floor(max(ay,by)/50)+1):
                    grid.setdefault((x,y),[]).append(edge)
                    edge_count+=1
                    if edge_count>500000:raise ValueError('Kartområdet innehåller för mycket data. Markera underlag manuellt.')
    spans=[]
    for a,b in zip(points,points[1:]):
        if a[4]!=b[4] or b[3]<=a[3]:continue
        n=max(1,math.ceil((b[3]-a[3])/35))
        for i in range(n):
            t=(i+.5)/n
            x,y=xy([a[0]+(b[0]-a[0])*t,a[1]+(b[1]-a[1])*t])
            candidates={}
            gx,gy=math.floor(x/50),math.floor(y/50)
            for dx in (-1,0,1):
                for dy in (-1,0,1):
                    for ax,ay,bx,by,k in grid.get((gx+dx,gy+dy),[]):
                        vx,vy=bx-ax,by-ay
                        f=max(0,min(1,((x-ax)*vx+(y-ay)*vy)/max(.0001,vx*vx+vy*vy)))
                        d=math.hypot(x-ax-f*vx,y-ay-f*vy)
                        if d<=18:candidates[k]=min(d,candidates.get(k,999))
            ranked=sorted(candidates.items(),key=lambda v:v[1])
            k=ranked[0][0] if ranked and (len(ranked)==1 or ranked[1][1]-ranked[0][1]>6) else 'unknown'
            start=a[3]+(b[3]-a[3])*i/n;end=a[3]+(b[3]-a[3])*(i+1)/n
            if spans and spans[-1]['kind']==k:spans[-1]['end']=end
            else:spans.append({'start':start,'end':end,'kind':k,'source':'osm' if k!='unknown' else 'unknown'})
    return spans

def fetch_suggestions(points):
    south,north=min(p[0] for p in points)-.001,max(p[0] for p in points)+.001
    west,east=min(p[1] for p in points)-.002,max(p[1] for p in points)+.002
    area=distance([south,west],[north,west])*distance([south,west],[south,east])/1e6
    if area>1200 or points[-1][3]>200000:
        raise ValueError('Automatiska förslag stöder rutter upp till 200 km i ett kartområde på högst 1 200 km². Markera underlag manuellt för denna rutt.')
    query=f'[out:json][timeout:30][maxsize:33554432];way["highway"]["surface"]({south},{west},{north},{east});out tags geom;'
    request=Request('https://overpass-api.de/api/interpreter',data=urlencode({'data':query}).encode(),
                    headers={'User-Agent':'AlvinsRuttbank/1.0','Content-Type':'application/x-www-form-urlencoded'})
    try:
        with urlopen(request,timeout=40) as response:
            raw=response.read(12*1024*1024+1)
        if len(raw)>12*1024*1024:raise ValueError()
        data=json.loads(raw)
        if data.get('remark'):raise ValueError()
    except Exception:
        raise ValueError('Karttjänsten kunde inte svara just nu. Inga underlag har ändrats. Försök senare eller markera manuellt.')
    return infer(points,data.get('elements',[]))
